Hw1

VIXER13 opened the pull request.

VIXER13 committed:
done

VIXER13 committed:
Заменил switch на if/else if

a-badin commented on CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+
+project(Process)
+
+set(SOURCES src/main.cpp src/process.cpp src/simple_dialog.cpp)
```
Предлагаю вынести process.hpp и process.cpp в отдельную библиотеку в папке lib

smirnov-vs commented on include/process.hpp:
```diff
@@ -0,0 +1,41 @@
+#include <exception>
+
+class Process {
+    std::array<int, 2> fd = {-1, -1};
```
лучше именовать переменные, разделить на два отдельных поля

VIXER13 replied:
То есть вместо массива использовать структуру, которую в случае необходимости представлять как массив? А не возникнет ли проблем с каким-нибудь выравниванием? Не лучше ли тогда создать внутри класса enum {IN, OUT}, и вместо индексов подставлять константы из него?

smirnov-vs replied:
почему мы здесь боремся за выравнивание?)
выравнивание двух элементов в массиве не отличается от выравнивания двух последовательных полей структуры

конкретно с enum будет просто больше кода, не вижу особого смысла в этом, писать надо проще

VIXER13 replied:
Прислушался к совету, завёл структуру.

smirnov-vs replied:
зачем структура? говорили просто про 2 отдельных поля

VIXER13 replied:
Мне кажется такой вариант
```
struct {
    int read = -1, write = -1;
} fd_in, fd_out;
```
выглядит лучше, чем такой
`int read_in = -1, write_in = -1, read_out = -1, write_out = -1;`
Потому что он позволяет выделить особенность, что это две связанные переменные и с ними надо обращаться, как с единым объектом.

smirnov-vs replied:
да, просто в конкретном случае классу достаточно знать всего 2 дескриптора, которые проще описать двумя переменными

VIXER13 replied:
Аа, понял, нам же не нужны read_in и write_out. Но ток у меня при их закрытии почему-то программа падает при попытке писать, например, в /bin/date. И не выдаёт никаких ошибок, просто завершается. Конечно, я понимаю, что это бессмысленно, но не понимаю, должна ли она при этом падать и если не должна, то как ловить? С /bin/cat всё работает.

smirnov-vs replied:
программа может упасть от сигнала, может упасть от исключения через `abort()`, скорее всего у тебя SIGPIPE, пишешь в закрытый пайп, добро пожаловать в дебаг

VIXER13 replied:
Убрал лишние дескрипторы, нужно ли ловить SIGPIPE и как это вообще делать?

smirnov-vs replied:
sigpipe можешь игнорировать приложением, которое использует бибилиотеку

a-badin commented on include/process.hpp:
```diff
@@ -0,0 +1,41 @@
+        message(str) {}
+
+    const char* what() const noexcept override {
+        return message.data();
```
у data может не оказаться 0 в конце строки. c_str() это гарантирует

a-badin replied:
Тут я не прав, с 11 гарантируется

a-badin commented on include/process.hpp:
```diff
@@ -0,0 +1,41 @@
+#include <string>
+#include <array>
+#include <exception>
+
```
Если будет библиотека, желательно добавить пространство имен

smirnov-vs commented on src/process.cpp:
```diff
@@ -0,0 +1,86 @@
+    if (pid < 0) {
+        throw ProcessException{"Fork failed"};
+    } else if (!pid) {
+        if (dup2(fd[0], 0) < 0 || dup2(fd[1], 1) < 0) {
```
стоит использовать константы типа `STDIN_FILENO`

smirnov-vs replied:
что-то не так, дочерний процесс пишет сам себе на stdin?
надо разобраться и протестировать

VIXER13 replied:
Про константу не знал, обязательно заменю, а про то кто куда пишет буду завтра разбираться.

a-badin commented on include/process.hpp:
```diff
@@ -0,0 +1,41 @@
+    }
+};
+
+#endif
```
Хорошим стилем считается добавлять комментарий какой ifdef закрывается

a-badin commented on src/main.cpp:
```diff
@@ -0,0 +1,21 @@
+            }
+        }
+    } else {
+        std::cerr << "Argc < 2." << std::endl;
```
наверное, надо вернуть код отличный от 0

smirnov-vs commented on src/process.cpp:
```diff
@@ -0,0 +1,86 @@
+bool Process::isReadable() const {
+    struct stat buff = {};
+    fstat(fd[0], &buff);
+    return buff.st_blksize;
```
проверка конца данных из дескриптора проверяется через `read`

smirnov-vs commented on src/process.cpp:
```diff
@@ -0,0 +1,86 @@
+}
+
+size_t Process::read(void* data, size_t len) {
+    return ::read(fd[0], data, len);
```
системный read возвращает не size_t

VIXER13 replied:
Да, он возвращает ssize_t, то есть тип той же размерности, что и size_t, но со знаком. Но я не вижу здесь смысла в явной конвертации типа, потому что беззнаковый тип и так считается шире, чем знаковый, поэтому он совершенно легально сконвертируется в нужный нам size_t. Но только мне всё равно не очень нравится то, что в случае ошибки нам вернётся очень огромное число. С другой стороны можно вспомнить про std::string::npos, который в свою очередь и является size_t(-1), тогда моя реализация будет повторять это поведение.

Или я неправильно понял претензию?

smirnov-vs replied:
проблему нашел, а претензию не понял
нужно обработать исключительный случай с -1 и дальше конвертировать в `size_t`

VIXER13 replied:
Теперь, и read, и write кидают исключения. Хотя, как мне кажется, это немного лишнее, учитывая, что функции что-то, да возвращают и size_t(-1) можно обработать как ошибку, в отличие от Exact методов, где исключения оправданы.

smirnov-vs replied:
ошибки отдельно удобно обрабатывать, результаты отдельно

smirnov-vs commented on src/process.cpp:
```diff
@@ -0,0 +1,86 @@
+    size_t readed = 0;
+    while (true) {
+        ssize_t read_out = ::read(fd[0], static_cast<int8_t*>(data) + readed, len - readed);
+        if (read_out == -1) {
```
это логика метода read, который можно тут использовать

smirnov-vs commented on src/simple_dialog.cpp:
```diff
@@ -0,0 +1,63 @@
+        return false;
+    } else if (choice == 1) {
+        std::cout << "Write string:";
+        std::cin >> str.data();
```
что если больше 256?

VIXER13 replied:
Ничего хорошего, заменю на std::string

smirnov-vs commented on src/simple_dialog.cpp:
```diff
@@ -0,0 +1,63 @@
+    } else if (choice == 3) {
+        std::cout << "Write string:";
+        std::cin >> str.data();
+        proc.writeExact(str.data(), strlen(str.data()));
```
надо переделать str на подходящий контейнер

VIXER13 replied:
Заменю на std::string

a-badin commented on src/process.cpp:
```diff
@@ -0,0 +1,86 @@
+            throw ProcessException{"Dup2 failed"};
+        }
+
+        if (execl(path.data(), path.data(), nullptr) < 0) {
```
c_str

smirnov-vs commented on CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+set(HEADERS include)
+
+include_directories(${HEADERS})
+add_executable(hw1 ${SOURCES})
```
где библиотека?)

smirnov-vs commented on src/main.cpp:
```diff
@@ -0,0 +1,21 @@
+
+        while (true) {
+            try {
+            if (!simpleDialog(proc))
```
надо отступы пофиксить

VIXER13 replied:
Действительно. Странно, что линтеры не ругались

smirnov-vs commented on src/main.cpp:
```diff
@@ -0,0 +1,21 @@
+
+int main(int argc, char **argv) {
+    if (argc > 1) {
+        Process proc(argv[1]);
```
там исключения внутри

a-badin commented on src/process.cpp:
```diff
@@ -0,0 +1,86 @@
+#include <sys/stat.h>
+
+Process::Process(const std::string& path) {
+    if (pipe(fd.data()) < 0) {
```
Одним пайпом тут не справится. С одним пайпом ваша программа работает до тех пор, пока ваш процесс повторяет то что принимает. Если, например, передать аргумент программу /bin/date, и передать ей строку 123, а потом считать 35 байт, то получим дату + 123. Но дочерний процесс не отправлял 123. Так получается потому, что и родитель и дочений процессы пишут и читают из одного пайпа.

a-badin reviewed (commented):
Правим второй пайп, выделяем библиотеку в отдельный  файл. Можно сделать свой echo процесс, чтобы на нем отлаживаться

VIXER13 committed:
Переместил модуль process в папку lib, остальной проект поместил в папку project.
В Process сделал два пайпа.

VIXER13 commented:
Я так и не понял как при помощи read проверять isReadable. Я если пытаюсь читать из пустого пайпа, то у меня всё повисает и ждёт, пока он чем-то заполнится.

VIXER13 committed:
Завёл переменную readable, которая следит за концом пайпа.

smirnov-vs commented:
Читать не надо, если пайп уже кончился, просто по факту окончания пайпа запоминать это

VIXER13 committed:
Заменил массивы для пайпа на структуры.
Методы read и write теперь кидают исключения в случае возврата -1.
Соответствующие изменения во всём коде.
Небольшие правки, чтобы линтеры не ругались.

VIXER13 committed:
Убрал лишние дескрипторы из объекта.

smirnov-vs commented on lib/src/process.cpp:
```diff
@@ -0,0 +1,112 @@
+size_t Process::read(void* data, size_t len) {
+    ssize_t temp = ::read(read_out, data, len);
+    if (temp == -1) {
+        readable = false;
```
это неверная проверка на EOF, гугл тут поможет

smirnov-vs commented on lib/src/process.cpp:
```diff
@@ -0,0 +1,112 @@
+            closeAllPipes(fd_in, fd_out);
+            throw ProcessException{"Dup2 failed"};
+        }
+        closeAllPipes(fd_in, fd_out);
```
может лучше применить RAII?
мне например тут и ниже довольно сложно выследить всегда ли ты все закрыл при ошибке

VIXER13 replied:
Добавил класс Descriptor

smirnov-vs commented on lib/src/process.cpp:
```diff
@@ -0,0 +1,112 @@
+        throw ProcessException{"Read error"};
+    }
+    for (size_t i = 0; i < len; ++i) {
+        if (static_cast<char*>(data)[i] == '\0') {
```
терминальный нуль это не EOF для read

VIXER13 replied:
Я просто не совсем правильно понял, что от меня в итоге требовалось. Я в итоге здесь не терминальный нуль поставил, а макрос EOF

smirnov-vs replied:
это все равно неверно, EOF это просто -1
что если я отправлю в пайп среди своих данных -1?

smirnov-vs commented on CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+set(LIB_HEADERS lib/include)
+
+include_directories(${HEADERS} ${LIB_HEADERS})
+add_executable(hw1 ${SOURCES} ${LIB_SOURCES})
```
нет библиотеки все еще тут

VIXER13 replied:
Теперь есть :)

VIXER13 committed:
Добавил класс дескриптора.
Выделил процесс в библиотеку в CMakeLists.txt.
Соответствующие правки.

smirnov-vs commented on process_lib/src/process.cpp:
```diff
@@ -0,0 +1,98 @@
+
+Process::Process(const std::string& path) {
+    std::array<Descriptor, 2> fd_in, fd_out;
+    if (pipe(reinterpret_cast<int*>(fd_in.data())) < 0 || pipe(reinterpret_cast<int*>(fd_out.data())) < 0) {
```
лучше не хачить, а сделать просто класс пайпу

smirnov-vs reviewed (changes requested):
еще остались проблемы

VIXER13 committed:
Добавил класс Pipe.
Исправил ошибку с определением EOF.

smirnov-vs commented on process_lib/src/descriptor.cpp:
```diff
@@ -0,0 +1,31 @@
+
+void Descriptor::move(Descriptor&& other) noexcept {
+    fd_ = other.fd_;
+    other.fd_ = -1;
```
а зачем выделять в отдельную функцию?

VIXER13 replied:
Чтобы не повторять код в конструкторе перемещения и в операторе присваивания с перемещением

smirnov-vs replied:
обычно это решают через copy-and-swap

VIXER13 replied:
Здесь, кстати, в другом даже ошибка была. Я не закрывал дескриптор other

VIXER13 replied:
В смысле, наоборот, не закрывал текущий дескриптор, прежде, чем его заменить

smirnov-vs commented on process_lib/src/pipe.cpp:
```diff
@@ -0,0 +1,33 @@
+    if(pipe(fds.data()) < 0) {
+        throw ProcessException{"Pipes failed"};
+    }
+    fds_[0] = std::move(Descriptor(fds[0]));
```
это и так rvalue

VIXER13 replied:
Да, но если я убираю мув, то у меня компилятор ругается

smirnov-vs replied:
что пишет? все должно быть ок

VIXER13 replied:
Ошибся с местом использования std::move, тут действительно всё хорошо

smirnov-vs commented on process_lib/src/pipe.cpp:
```diff
@@ -0,0 +1,33 @@
+}
+
+void Pipe::move(Pipe&& other) noexcept {
+    fds_[0] = std::move(other.fds_[0]);
```
мув инта

VIXER13 replied:
Это оправдано тем, что при муве у нас в other.fds_[0] внутренний int будет заменён на -1, чтобы дескриптор не закрылся при выходе из зоны видимости объекта.

smirnov-vs commented on process_lib/src/pipe.cpp:
```diff
@@ -0,0 +1,33 @@
+}
+
+Descriptor& Pipe::getDescriptor(const size_t i) noexcept {
+    return fds_[i];
```
тут скорее доступ через enum нужен, только 2 варианта

smirnov-vs commented on process_lib/src/process.cpp:
```diff
@@ -0,0 +1,91 @@
+}
+
+void Process::closeStdin() {
+    ::close(write_in);
```
это класс, так что лучше вызывать его метод close, который уже делает всё

VIXER13 committed:
Добавил в Pipe enum из двух значений.
В Descriptor добавил функцию close
